pkg/plugins/manager/loader/loader.go

```diff
@@ -75,15 +82,75 @@ func (l *Loader) Load(ctx context.Context, src plugins.PluginSource) ([]*plugins
 	return l.loadPlugins(ctx, src, found)
 }
 
+func (l *Loader) getPublicKey(features plugins.FeatureToggles) (string, error) {
+	// The fallback public key is used when the feature flag is not enabled.
+	publicKeyText := `-----BEGIN PGP PUBLIC KEY BLOCK-----
+Version: OpenPGP.js v4.10.1
+Comment: https://openpgpjs.org
+	
+xpMEXpTXXxMFK4EEACMEIwQBiOUQhvGbDLvndE0fEXaR0908wXzPGFpf0P0Z
+HJ06tsq+0higIYHp7WTNJVEZtcwoYLcPRGaa9OQqbUU63BEyZdgAkPTz3RFd
+5+TkDWZizDcaVFhzbDd500yTwexrpIrdInwC/jrgs7Zy/15h8KA59XXUkdmT
+YB6TR+OA9RKME+dCJozNGUdyYWZhbmEgPGVuZ0BncmFmYW5hLmNvbT7CvAQQ
+EwoAIAUCXpTXXwYLCQcIAwIEFQgKAgQWAgEAAhkBAhsDAh4BAAoJEH5NDGpw
+iGbnaWoCCQGQ3SQnCkRWrG6XrMkXOKfDTX2ow9fuoErN46BeKmLM4f1EkDZQ
+Tpq3SE8+My8B5BIH3SOcBeKzi3S57JHGBdFA+wIJAYWMrJNIvw8GeXne+oUo
+NzzACdvfqXAZEp/HFMQhCKfEoWGJE8d2YmwY2+3GufVRTI5lQnZOHLE8L/Vc
+1S5MXESjzpcEXpTXXxIFK4EEACMEIwQBtHX/SD5Qm3v4V92qpaIZQgtTX0sT
+cFPjYWAHqsQ1iENrYN/vg1wU3ADlYATvydOQYvkTyT/tbDvx2Fse8PL84MQA
+YKKQ6AJ3gLVvmeouZdU03YoV4MYaT8KbnJUkZQZkqdz2riOlySNI9CG3oYmv
+omjUAtzgAgnCcurfGLZkkMxlmY8DAQoJwqQEGBMKAAkFAl6U118CGwwACgkQ
+fk0ManCIZuc0jAIJAVw2xdLr4ZQqPUhubrUyFcqlWoW8dQoQagwO8s8ubmby
+KuLA9FWJkfuuRQr+O9gHkDVCez3aism7zmJBqIOi38aNAgjJ3bo6leSS2jR/
+x5NqiKVi83tiXDPncDQYPymOnMhW0l7CVA7wj75HrFvvlRI/4MArlbsZ2tBn
+N1c5v9v/4h6qeA==
+=DNbR
+-----END PGP PUBLIC KEY BLOCK-----
+`
+	if !features.IsEnabled("pluginsAPIManifestKey") {
+		return publicKeyText, nil
+	}
+
+	var data struct {
+		Items []struct {
+			KeyID  string `json:"keyId"`
+			Since  int64  `json:"since"`
+			Public string `json:"public"`
+		}
+	}
+
+	resp, err := http.Get(l.grafanaURL + "/api/plugins/ci/keys")
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
+		return "", err
+	}
+
+	if len(data.Items) == 0 {
+		return "", errors.New("missing public key")
+	}
+
+	return data.Items[0].Public, nil
+}
+
 func (l *Loader) loadPlugins(ctx context.Context, src plugins.PluginSource, found []*plugins.FoundBundle) ([]*plugins.Plugin, error) {
 	var loadedPlugins []*plugins.Plugin
+
+	publicKeyText, err := l.getPublicKey(l.features)
+	if err != nil {
+		return nil, fmt.Errorf("could not get public key: %w", err)
+	}
+
 	for _, p := range found {
 		if _, exists := l.pluginRegistry.Plugin(ctx, p.Primary.JSONData.ID); exists {
 			l.log.Warn("Skipping plugin loading as it's a duplicate", "pluginID", p.Primary.JSONData.ID)
 			continue
 		}
 
-		sig, err := signature.Calculate(ctx, l.log, src, p.Primary)
+		sig, err := signature.Calculate(ctx, l.log, src, p.Primary, publicKeyText)
```

Comment:
One of my initial thoughts when looking through this is that maybe it makes sense to try encapsulate this logic inside the `signature` package instead. It feels a little weird for the loader to fetch the key and provide it. That would likely mean we'll have to make a Wire service here, but it would make for better encapsulation IMO

Reply:
I moved it to its own package (`manifestverifier`), but it's still under `loader` because I need to expose the grafana URL there for the tests. If I move it to the configuration it may not be needed and everything could be under `signature` :thought_balloon:. Let me try that :+1:

Reply:
Sweet!